### ISSUE TITLE
Avoid external solver when evaluating functions

### ISSUE THREAD

**Issue by yzhang90** [kframework/kore#1318]
I'm now debugging the performance issue of `totalSupply-spec.k`. in the spec, it calls #parseByteStack on a very long hex string (around 14,000 chars) and there is only one top level predicate, which is `requires 0 <=Int ACCT_ID    andBool ACCT_ID    <Int (2 ^Int 160)`
In the log, I found that
```
(assert (and <0> (= (and (<= 0 <1> ) (< <1> 1461501637330902918203684832716283019655932542976 ) ) true ) ) )
```
is sent to Z3 ~7000 times. it matches the number of times that #parseHexBytes is called (the function is called on every two chars).
I guess the top level predicate is evaluated whenever a function is called. However, if the function arguments are concrete, there is no need to evaluate top level predicate at all.

**Comment by ttuegel** [comment]
At present we use SMT to check satisfiability of `P ∧ R` where `P` is the top-level condition and `R` is the requires clause. Instead, we should check unsatisfiability of `P ∧ ¬ R` (requires #1065). When the function's arguments are constructor-like (*not* only concrete) we can evaluate `¬ R` without using the external solver.

**Comment by yzhang90** [comment]
I think this is good to have anyway. However, in some (common) cases, this is not enough. for example,
```
rule f(x) => g(x) requires x >5
```
current term is `f(3)` and top level condition is P (e.g., V > 0), we will then check unsat of `P /\ ¬(3 > 5)` and we'll send P to SMT, which is unnecessary.
I think a more general question is that `can we tell from the syntax of the arguments if top-level condition is useful for evaluating the function`?
We can analyze most common cases and add heuristic for that.

**Comment by virgil-serbanuta** [comment]
Wait, why wouldn't `3>5` be fully simplified inside the haskell code?

And, if the `requires` clause is `top` or `bottom`, we normally skip sending it to the SMT:
https://github.com/kframework/kore/blob/master/kore/src/Kore/Step/SMT/Evaluator.hs#L87
If that does not happen for fuction evaluation, then we should fix it. I.e. hopefully we only send something to the SMT when there is a local condition whose validity might depend on the top level one.

So, what am I missing?

**Comment by yzhang90** [comment]
Let me reuse my example.
```
rule f(x) => g(x) requires x >5
```
Currently function to evaluate is f(6) and top-level predicate is P.
If i understand correctly, when applying the above rule, we will check the sat of the following:
`Ceil(f(x) /\ f(6)) /\ x > 5 /\ P`, and it becomes `true /\ P`, which we need to send to Z3.
What Tom proposed is to check the unsat of `Ceil(f(x) /\ f(6)) /\ ¬(x > 5) /\ P`(which says the top level predicate implies the requires clause), and it becomes `false /\ P`, which is trivially unsat.

However, checking unsat will fail on evaluating f(3), because `P /\ ¬(3 > 5)` is still `P` and we need to consult SMT to determine its satisfiability.

**Comment by ttuegel** [comment]
> hopefully we only send something to the SMT when there is a local condition whose validity might depend on the top level one.

We do not do that, we only check that `P ∧ R` is satisfiable, so if `R` is `\bottom` we should not send anything` but we will send when `R` is `\top` if `P` is not also `\top`. That's why I suggested that for functions we check that `P ∧ ¬R` is `unsat` instead. Actually, I just realized we need to check all of these:

1. If `R` is `\bottom`, the rule does not apply.
1. If `¬R` is `\bottom`, we immediately infer that the rule applies.
1. Otherwise, we check that `P ∧ ¬R` is `unsat`, that is, check that `P → R`, in which case we infer that the rule applies.

Only the last check needs SMT. All of these need #1065 first.
